design-documents/storefront/pricing.md

```diff
@@ -0,0 +1,311 @@
+
+The impact form cases describe above will be doubled(or even tripled) if we introduce a new storefront service with existing index structure inside. Thus, we need some other way to work with prices in storefront.
+
+## Glossary
+- Customer group - allocate each customer to a group and control store behaviour(catalog restrictions, pricing, discounts, B2B, CMS, payments, shipping, etc) according to which group a customer belongs to  
+- Website, Store View - abstract Magento scopes - https://docs.magento.com/user-guide/configuration/scope.html . Magento modules and third-party extensions may utilize this scope as they want.
+- Base price - the initial product price. Different values could be set per website in Magento monolith.
+- Tier price - have two meanings `customer-specific price` and `volume based price`.  
+- Special price - time based price for the product. Original price is strikedout in the UI. (e.g. was ~~$100.00~~, now $99.00) 
+- Catalog Rule - Magento functionality that allows to apply the same discount to multiple products
+
+
+## Goals
+
+- Support up to 5,000,000 products and 15,000 customer groups
+- Establish efficient sync of prices between Magento monolith and Magento storefront
+- Reduce the size of pricing index
+- Provide reliable support for personalized prices
+
+## Solution
+
+All price dimensions aren't used exclusively for pricing: 
+multiple `websites` may have the same price, but represent different web domains; `customer group` could represent the company in b2b scenarios or class of customer service; `product` may have different pricing on one website only.
+The solution is to reuse data where possible and make separation of dimensions, so `websites` or `customer groups` which are not a part of pricing will not trigger `price index` explosion.
+
+### Price books
+
+The `price book`(price list) is a new entity that holds a list of product's prices for a sub-set of catalog. 
+The `price book` is a non-scoped entity, but it may hold information about linked customers, websites, etc. 
+Instead of direct lookup in price index by customer_group, website and product, we will detect the customer's price book first, then we will extract two price books from the index: default and current `price book`. 
+The resulting product price will be the value from current price book if it's exists, otherwise the product price will be extracted from default price book:
+
+![Price books diagram](pricing/pricebooks.png)
+
+Guidelines:
+* There is no need to resolve price book on each HTTP call. Resolved price book could be stored in JWT during "login" step 
```

Comment:
Can you explain this more - what specific data is going into a JWT?

Reply:
Just in case, here is a doc about jwt - https://github.com/magento/architecture/blob/master/design-documents/service-isolation/detailed-architecture.md

I didn't want to cover this in the proposal because it could be potentially unified with other data in the customer context. However, I see the following options:
* Put price book id (and probably default price book id) in the customer context. Pass this context to all services with each request.
* Store price book ids  in customer context, but extract price book ids on GraphQL side and pass them as arguments to pricing/catalog/search services
* Pass customer_id only to the services. Each service may extract price book ids using separate network call to CPQ service.

I don't like option #3 because of relatively slow performance. #1 and #2 are almost identical to me, but option #1 seems to be more flexible and reusable for other scenarios. However, #1 moves more data with each client->graphl and graphql->services requests.